include/fi_tagged_rx.h

```diff
@@ -0,0 +1,166 @@
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _FI_TAGGED_RX_H_
+#define _FI_TAGGED_RX_H_
+
+#include <fi.h>
+#include <fi_mem.h>
+#include <rdma/fi_tagged.h>
+
+/*
+ * Tagged recv list template: provides a set of utility api's for
+ * software-level tag-matching and managing posted/unexpected
+ * messages.
+ * 
+ * `entrytype' structure should have (but not limited to) the
+ * following fields:
+ *
+ *         - struct dlist_entry entry;
+ *         - struct fi_msg_tagged tmsg;
+ *         - int index;
+ *
+ * Interfaces:
+ *
+ * <name>_create - Create tagged-recv list, with a maximum of
+ * `num_rx' recv entries in freestack
+ *
+ * <name>_free - Frees the tagged-recv list
+ *
+ * <name>_entry_alloc - gets an entry from freestack. Caller can insert
+ * it into the posted/unexpected list as needed
+ *
+ * <name>_search_post - search for a matching recv entry with the given
+ * tag/addr in posted list
+ * 
+ * <name>_search_unexp - search for a matching recv entry with the
+ * given tag/match/addr in unexpected list
+ */
+#define DECLARE_TAGGED_RX(entrytype, name)				\
+DECLARE_FREESTACK(entrytype, name ## _fs)				\
+struct name {								\
+	struct dlist_entry post_list;					\
+	struct dlist_entry unexp_list;					\
+	struct name ##_fs *fs;						\
+};									\
+									\
+static inline struct name *name ## _create(ssize_t num_rx)		\
+{									\
+	int idx, i;							\
+	entrytype *entry;						\
+	struct name *list;						\
+	struct dlist_entry *item;					\
+	list = calloc(1, sizeof(struct name));				\
+	if (!list)							\
+		return NULL;						\
+									\
+	list->fs = name ## _fs_create(num_rx);				\
+	if (!list->fs) {						\
+		free(list);						\
+		return NULL;						\
+	}								\
+									\
+	dlist_init(&list->post_list);					\
+	dlist_init(&list->unexp_list);					\
+	for (i = 0; i < num_rx; i++) {					\
+		idx = freestack_pop(list->fs);				\
+		assert(idx != FREESTACK_EMPTY);				\
+		entry = &list->fs->buf[idx];				\
+		entry->index = idx;					\
+		dlist_insert_tail(&entry->entry, &list->post_list);	\
+	}								\
+	for (i = 0; i < num_rx; i++) {					\
+		item = list->post_list.next;				\
+		dlist_remove(item);					\
+		entry = container_of(item, entrytype, entry);		\
+		freestack_push(list->fs, entry->index);			\
+	}								\
```

Comment:
I was just trying to use the freestack interfaces, rather than using the internal array. So, if freestack implementation changes later, this does not need any change.

But, agree that it will be much simpler to directly access the array :)